[5.8]  Make the redis queue blocking pop atomic

matt-allan opened the pull request:
## Overview

This PR changes the redis queue driver's blocking pop implementation to pop jobs atomically, fixing #22939.  The pattern is [explained in the Redis docs](https://redis.io/commands/blpop#pattern-event-notification).

When we push a job we also push a 'notification key'.  The notification list is named using the pattern `queues:{{name}}:notify`.

```
MULTI
RPUSH queues:default {{payload}}
RPUSH queues:default:notify 1
EXEC
```

When the worker attempts to reserve the job it first tries a blocking pop on the notification list.  As soon as a key is aded to the notification list the worker returns and then attempts to pop a job like normal, using the existing Lua script.

```
BLPOP queues:default:notify {{block_for}}
EVAL {{pop_script}} 2 queues:default queues:default:reserved {{available_at}}
```

There is a small chance the worker might crash between popping the notification element and popping the job onto the reserved queue. That is ok - once the worker restarts it will block for `block_for` seconds, then attempt to pop the job again.

## Upgrade Notes

If you upgrade the workers before upgrading the PHP processes pushing the jobs the worker will block waiting for a notification key that is never set.  This means you won't be able to mix pre 5.8 producers with 5.8+ workers without slowing down the workers considerably.

If for some reason you have to upgrade the workers first you can set `block_for` to `null` until you are fully upgraded.

Commit:
Make the redis queue blocking pop atomic

Use a separate list `queues:{{name}}:notify` to allow us to use the
BLPOP command without risking lost jobs.

taylorotwell commented:
Probably would need some review from @halaei. I'm not super thrilled about introducing more breaking changes.

halaei commented:
@taylorotwell it seems to be good at first glance. Needs deeper investigation.

I guess the idea is somehow similar to my former [pull request](https://github.com/laravel/framework/pull/23057) in the sense that it requires adding an extra key per queue. The advantage is that @yuloh's solution doesn't require reversing the order of queues, which is nice.
Please give me some time to check some edge-cases as well.

It will also be nice if you can give feedback on my Redis module: https://github.com/halaei/lqrm

halaei commented on src/Illuminate/Queue/RedisQueue.php:
```diff
@@ -88,7 +88,16 @@ public function size($queue = null)
+        $this->getConnection()->rpush($this->getQueue($queue).':notify', '1');
+        $id = $this->pushRaw($this->createPayload($job, $this->getQueue($queue), $data), $queue);
+        $this->getConnection()->exec();
+
```
`rpush queue:notify 1` is missing from  LuaScripts::migrateExpiredJobs().
Also I think this code is better to be moved to LuaScript as well.

halaei commented:
Other than the required changes, the drawbacks I can think of are:

1. O(1) extra memory & cpu cost per job.
2. A chance of a redundant wait in case of a crash.

Both are practically negligible.

Commit:
Notify when migrating expired jobs

Commit:
Use a Lua script for RedisQueue::push

halaei commented:
My suggestion is not good enough. It doesn't work fine in case of worker crash. Although, it can be fixed by lua scripts.

yuloh commented:
> To fix issue 1 (workers waiting for notify when there is no notify but a job is ready), let the blocking workers pop notify after they try pop from the main queue. And block for notify only if there was no job available:

I considered this but it doesn't completely fix the problem.  If the worker checks the queue and a job isn't ready, it will still block for N seconds before it checks the queue again.  So it prevents the wait if a job is ready when the worker first checks the queue but it doesn't help if the job is pushed after the worker begins blocking.

> To fix issue 2 (memory leak in case of blocking push and non-blocking pop), let non-blocking workers as well pop from notify queue

This seems fine to me.  It still doesn't solve the problem if pre 5.8 workers are interacting with 5.8+ clients though.  The pre 5.8 worker won't be popping from the notify queue so the memory leak would still exist.  It would only solve the problem of 5.8+ non blocking workers working alongside 5.8+ blocking clients.

Commit:
Add comments to the Lua scripts

Commit:
Resize the redis queue notifications list

If a worker crashes after it's popped a notification but before it's
popped the job the next worker that attempts to pop the job will block
waiting for the notification unnecessarily.  To prevent this we can
check that all ready jobs have a notification and if they don't create
any missing notifications.

Similarly if a worker does not pop the notification for some reason
(possible if we are mixing workers from different versions) the notify
list can grow too long.  To prevent this we can trim the notification
list to ensure the # of notifications matches the # of ready jobs.

yuloh commented:
If you are not already using blocking pop you don't need to do anything.  You can upgrade like you normally would as all blocking pop functionality will be disabled.

If you _are_ using the blocking pop functionality you should restart your workers before restarting the web server.  This is to prevent the clients from pushing notifications the workers are not removing, which could cause the Redis memory usage to go up temporarily until you start a 5.8 worker.

The memory usage is really small (~.19MB for 100k jobs) and the excess notifications will go away as soon as you start a 5.8+ blocking pop worker.  Even on the smallest server you would have to push millions of jobs before the memory usage became a problem 😆

yuloh commented:
@halaei I just pushed a change that should address most of your concerns.

Before popping the job we check if the notify list (`queues:default:notify`) is the same length as the ready queue (`queues:default`).  If the notify list is too long we trim it.  If it's too short we push the missing elements.

Creating the missing notifications will prevent us from having to block unnecessarily when the notification is missing, which could happen if a 5.7 client pushed the job or a worker crashed between popping the notification and the job.  If a job is pushed without a notification _after_ a worker has already started blocking any other worker attempting to pop a job on that queue will be able to unblock the first worker.

Trimming the notification list will prevent a memory leak if the worker didn't pop the notification, which could happen if you were using 5.7 blocking pop workers with 5.8 blocking pop clients.

> If a process pushes jobs with notify, but workers are in non-blocking mode, the notify flags will be in memory for ever.

I didn't attempt to solve this issue (using non blocking workers with blocking clients).  If we wanted to solve this we would have to make all workers pop notifications, even if they aren't using blocking pop.  Even then we wouldn't be able to prevent someone from using a 5.8 blocking pop client with a 5.7 client which causes the same issue.  This seems like enough of an edge case I think we can ignore it?

halaei commented:
@yuloh The last commit might address some of the concerns, but I fee it doesn't worth the extra effort and this PR is good enough without it. Please let me try something else as well.

halaei commented:
@yuloh Please check https://github.com/halaei/framework/commit/44860e5d0c85eb30f4473d1c5be7a28961eb0997

yuloh commented:
> @yuloh The last commit might address some of the concerns, but I fee it doesn't worth the extra effort and this PR is good enough without it.

The cost is only a single call to Redis and two O(1) commands if the notify queue is the correct length.  The benefit is the workers rarely have to block unnecessarily and there will never be a memory leak.

> @yuloh Please check halaei@44860e5

I left a comment on your commit (https://github.com/halaei/framework/commit/44860e5d0c85eb30f4473d1c5be7a28961eb0997#r32089485).

If we make those changes every Redis queue user will need to make sure they restart the workers first, even if they aren't using blocking pop.  If they don't restart the workers first once they finally do restart them the memory usage will never go back down since the excess notifications will never be popped.  Ideally we could limit the breaking changes to only users that use blocking pop.

I'm fine with either solution.  The main difference is the resize script will prevent the possibility of a memory leak, but the memory leak would only ever happen if you mixed Laravel applications in a way that isn't supported.  If someone does that they can easily manually `DEL queues:default:notify` to resolve the excess memory.

halaei commented:
@yuloh Even if they don't manually `DEL queues:default:notify`, the notify list will eventually gets empty when there is no job to process.

> I'm fine with either solution.

Thanks. I'll send it as an alternative PR.
